bittide-instances/src/Bittide/Instances/Tests/FullMeshHwCc/IlaPlot.hs

```diff
@@ -112,6 +173,67 @@ accwindow _ clk rst ena =
            |     d'                  -> ((d,     s + 1),    NoChange)
            | otherwise               -> ((d,     s - 1),    NoChange)
 
+-- | Calculates the difference of a wrapping counter between two
+-- points in time taking potential overflows into account. The
+-- captured difference is wrapped into an option type, which defaults
+-- to 'Nothing' as soon as the difference exceeds the capacity of the
+-- returned index type.
+--
+-- The difference is measured against a stored reference, which is
+-- taken from the counter whenever the additional input signal gets
+-- asserted. 'Nothing' is output before the first assertion of this
+-- line.
+--
+-- The counter is assumed to only increase over time and may overflow
+-- several times until the next value gets stored. This means that the
+-- returned difference is measured against the point in time where the
+-- reference was stored (accumulating every overflow since then) and
+-- not against the actual value of the counter at that time. The
+-- counter can increase by any value in the range of the counter's
+-- type per cycle.
+overflowResistantDiff ::
+  forall dom n m.
+  (KnownDomain dom, KnownNat n, KnownNat m, 1 <= n, 1 <= m) =>
+  Clock dom ->
+  Reset dom ->
+  Enable dom ->
+  -- | Take the current counter value a the new reference if asserted
+  Signal dom Bool ->
+  -- | The input counter
+  Signal dom (Unsigned n) ->
+  -- | The measured difference as long as it fits into the output type
+  Signal dom (Maybe (Index m))
+overflowResistantDiff clk rst ena trg cnt =
+  mealyB clk rst ena transF Nothing (cnt, trg)
+ where
+  transF = maybe (\(c, s) -> if s then store c else (Nothing, Nothing)) $
+    \(refValue, overflows, lastDiff) (curValue, newRef) ->
+      if newRef then store curValue else
+        let diff = curValue - refValue
+            overflows'
+              | diff < lastDiff = satSucc SatError overflows
+              | otherwise       = overflows
+         in if    overflows  == maxOverflows && diff < lastDiff
+               || overflows' == maxOverflows && diff > natToNum @(Mod (m-1) (2^n))
+            then (Nothing, Nothing)
+            else ( Just (refValue, overflows', diff)
+                 , case subZeroMin @(BitSize (Index m)) @n of
+                     Dict -> case minLeq @(BitSize (Index m)) @n of
+                       Dict -> Just $ bitCoerce $ concatUnsigneds overflows'
+                         $ checkedTruncateB
+                             @(Min (BitSize (Index m)) n)
+                             @(n - Min (BitSize (Index m)) n)
+                             diff
+                 )
+
+  store = ( , Just 0) . Just . ( , 0, 0)
+
+  maxOverflows :: Unsigned (SubZero (BitSize (Index m)) n)
+  maxOverflows =
+    let x :: Unsigned (BitSize (Index m))
+        x = bitCoerce (maxBound :: Index m) `shiftR` natToNum @n
+     in case subZeroMin @(BitSize (Index m)) @n of
+          Dict -> checkedTruncateB x
```

Comment:
I think this is correct, but I can't quite convince myself due to this function doing multiple things at once. I think this should simplify quite a bit if we do a couple of things:

 * Make `overflow` overflows unrecoverable and reflect that in the type signature. I.e., the output type should be a three-state: `data CounterDiff m = Wait | Ok (Index m) | Error` (name as you see fit). I think this is sensible to do: if an output overflow happens, that's a system design failure in our context, right?
 * Split the function into two: one that assumes that its output is bigger or equal to its input: `(n <= m) => .. -> Signal dom (Maybe (Unsigned n)) -> Signal dom (Maybe (Unsigned m))`. I.e., this function should deal with _input_ overflows but not _output_ overflows. The other function can use that function to implement `.. -> Signal dom (Maybe (Unsigned n)) -> Signal dom (CounterDiff m)`